components/support/ffi/src/handle_map.rs

```diff
@@ -744,9 +744,14 @@ impl<T> ConcurrentHandleMap<T> {
     /// Note that this requires taking the map's write lock, and so it will
     /// block until all other threads have finished any read/write operations.
     pub fn delete(&self, h: Handle) -> Result<(), HandleError> {
-        // XXX figure out how to handle poison...
-        let mut map = self.map.write().unwrap();
-        map.delete(h)
+        // We use `remove` and not delete (and use the inner block) to ensure
+        // that if `v`'s destructor panics, we aren't holding the write lock
+        // when it happens, so that the map itself doesn't get poisoned.
+        let v = {
+            let mut map = self.map.write().unwrap();
+            map.remove(h)
+        };
+        v.map(drop)
```

Comment:
I guess this is to make it explicit that the value gets dropped?

Reply:
yeah, it's that or `v.map(|_| ())`, this is just more explicit.